compiler/examples/gimli/arm-m4/gimli.jazz

```diff
@@ -32,43 +27,57 @@ fn gimli (reg u32 state) {
 
     for round = N_ROUND downto 0 {
     for column = 0 to N_COLUMN {
-    x = (u32)[state + WORD_SIZE * (0 * WORD_SIZE + column)];
-    x = rotate(x, 24);
-    y = (u32)[state + WORD_SIZE * (1 * WORD_SIZE + column)];
-    y = rotate(y, 9);
-    z = (u32)[state + WORD_SIZE * (2 * WORD_SIZE + column)];
+    x = state[0 + column];
+    x <<r= 24;
+    y = state[1 + column];
+    y <<r= 9;
+    z = state[2 + column];
 
     a = x ^ (z << 1);
     b = y & z;
     a ^= b << 2;
-    (u32)[state + WORD_SIZE * (2 * WORD_SIZE + column)] = a;
+    state[2 + column] = a;
 
     a = y ^ x;
     b = x | z;
     a ^= b << 1;
-    (u32)[state + WORD_SIZE * (1 * WORD_SIZE + column)] = a;
+    state[1 + column] = a;
 
     a = z ^ y;
     b = x & y;
     a ^= b << 3;
-    (u32)[state + WORD_SIZE * (0 * WORD_SIZE + column)] = a;
+    state[0 + column] = a;
     }
 
     if (round % 4 == 0) {
-    swap(state, 0, 1);
-    swap(state, 2, 3);
+    state = swap(state, 0, 1);
+    state = swap(state, 2, 3);
     }
 
     if (round % 4 == 2) {
-    swap(state, 0, 2);
-    swap(state, 1, 3);
+    state = swap(state, 0, 2);
+    state = swap(state, 1, 3);
     }
 
     if (round % 4 == 0) {
-    a = (u32)[state + WORD_SIZE * 0];
+    a = state[0];
     b = rc | round;
     a ^= b;
-    (u32)[state + WORD_SIZE * 0] = a;
+    state[0] = a;
     }
     }
+    return state;
+}
+
+export
+fn gimli (reg u32 io) {
+  stack u32[12] state;
+  inline int i;
+  for i = 0 to 12 {
+    state[i] = (u32)[io + 4 * i];
+  }
+  state = gimli_ref(state);
+  for i = 0 to 12 {
+    (u32)[io + 4 * i] = state[i];
+  }
```

Comment:
We probably want
```
export
fn gimli (reg u32 io) {
  stack u32[12] state;
  inline int i;
  reg u32 tmp;

  for i = 0 to 12 {
    tmp = (u32)[io + 4 * i];
    state[i] = tmp;
  }

  state = gimli_ref(state);

  for i = 0 to 12 {
    tmp = state[i];
    (u32)[io + 4 * i] = tmp;
  }
}
```
?
Or should the compiler figure this out?

Reply:
We want the whole function to disappear and just export `gimli_ref`.

Reply:
I see. But this is not happening in either x86 or ARM, and in ARM the failure comes from moving between memory locations. At which point should the compiler realize this? Stack allocation?

Reply:
In x86, the intermediate register is introduced in lowering, which is too late IMHO.

What I’ve been dreaming of is a pass before lowering that introduces all required intermediate moves. This might be optional (enabled in “easy mode”, triggering warnings in “jasmin mode” , failing loudly in ”Peter’s mode”).

Reply:
Maybe we need to write all these dreams down at some point. Describe the ideal architecture of the compiler. I think also of the split of linearization. And specify precisely the modes and what they enforce.